Fix array casts as list type

rajyan opened the pull request:
fixes array cast of types as list

rajyan committed:
add tests for array casting

rajyan committed:
fix array cast of simple types as list

rajyan commented:
For numeric, falsy, non-empty string, it's weird that the casted type is `array<0, non-empty-string>` ~~(which is wrong)~~. I think there is another underlying bug.

rajyan commented:
ok, it was not wrong, but intersection of two constant arrays narrowed to array type was the cause.

rajyan committed:
fix array cast of accessory types

rajyan committed:
improve intersection of two ConstantArrayType

rajyan commented:
Very weird only failing on windows...

rajyan commented:
solved
the namespace was wrong

rajyan committed:
fix namespace of test

rajyan commented:
Noticed that failing tests of composer is caused from my previous PR https://github.com/phpstan/phpstan-src/commit/8b456c524c6dd8a782569c8c23e727189494ec32
I'll look into it.

rajyan commented:
I think the composer failure is caused because of this `composer update`
https://github.com/phpstan/phpstan/blob/c1ea106f9a101643d4e224c0917cef4eb0c1f60c/.github/workflows/integration-tests.yml#L127
The CI of composer is failing right now while migrating to PHPstan 1.9.0
https://github.com/composer/composer

I'm not sure why the composer integration test is running `composer update`.

staabm commented on tests/PHPStan/Rules/Functions/data/array-cast-list-types.php:
```diff
@@ -0,0 +1,25 @@
+foo((array) (fn () => 'closure'));
+foo((array) $nonEmptyString);
+foo((array) $nonFalsyString);
+foo((array) $numericString);
```
The test code should be wrapped into a method or function (no top level code, when possible as it would get executed when running the testsuite)

ondrejmirtes replied:
It's better to define parameter types as `@param` rather than `@var`.

rajyan committed:
fix tests to use method calls, param tags and move the test to CallMethodsRuleTest

rajyan commented:
fixed the tests and moved to CallMethodsRule
Playground result to show that the test is failing without this PR.
https://phpstan.org/r/e7dcb950-2cdc-4a07-9d4d-e89b58856b48

rajyan committed:
cs fix

rajyan committed:
fix test for literal string

rajyan commented:
I was taking literal-string wrong.
Fixed the test and the correct playground result is this
https://phpstan.org/r/75e6f175-55e5-42aa-9cfc-64343c705af8

ondrejmirtes commented:
Thank you!
